src/Aspire.Hosting/DistributedApplicationBuilder.cs

```diff
@@ -226,20 +233,62 @@ public DistributedApplicationBuilder(DistributedApplicationOptions options)
         }
 
         // Publishing support
-        _innerBuilder.Services.AddLifecycleHook<Http2TransportMutationHook>();
-        _innerBuilder.Services.AddLifecycleHook<DashboardManifestExclusionHook>();
+        Eventing.Subscribe<BeforeStartEvent>(MutateHttp2TransportAsync);
         _innerBuilder.Services.AddKeyedSingleton<IDistributedApplicationPublisher, ManifestPublisher>("manifest");
 
+        Eventing.Subscribe<BeforeStartEvent>(ExcludeDashboardFromManifestAsync);
+
         // Overwrite registry if override specified in options
         if (!string.IsNullOrEmpty(options.ContainerRegistryOverride))
         {
-            _innerBuilder.Services.AddLifecycleHook<ContainerRegistryHook>();
+            Eventing.Subscribe<BeforeStartEvent>((e, ct) => UpdateContainerRegistryAsync(options));
         }
 
         _innerBuilder.Services.AddSingleton(ExecutionContext);
         LogBuilderConstructed(this);
     }
 
+    private Task MutateHttp2TransportAsync(BeforeStartEvent beforeStartEvent, CancellationToken cancellationToken)
```

Comment:
nit: I'm not loving these being methods inside of this class.

Reply:
Yeah some of these global scenarios feel better as a lifecycle hook mostly for code structuring reasons. We can put these things into a separate class possibly with an external method but that seems like overkill in the same way that a lifecycle hook does.

Reply:
I've moved them out into a separate internal class (grouped together just for some structuring). Also made them static since they just rely on the event itself now for everything they need.